itest/multisig_test.go

```diff
@@ -0,0 +1,597 @@
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/commitment"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/proof"
+	"github.com/lightninglabs/taproot-assets/tappsbt"
+	"github.com/lightninglabs/taproot-assets/taprpc"
+	wrpc "github.com/lightninglabs/taproot-assets/taprpc/assetwalletrpc"
+	"github.com/lightninglabs/taproot-assets/taprpc/mintrpc"
+	"github.com/lightninglabs/taproot-assets/tapsend"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/lightningnetwork/lnd/lnrpc/signrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/walletrpc"
+	"github.com/lightningnetwork/lnd/lntest/node"
+	"github.com/stretchr/testify/require"
+)
+
+// testMultiSignature tests that we can use multi signature on all levels of the
+// Taproot Assets Protocol. This includes the BTC level, the asset level and the
+// group key level.
+func testMultiSignature(t *harnessTest) {
+	ctxb := context.Background()
+	ctxt, cancel := context.WithTimeout(ctxb, defaultWaitTimeout)
+	defer cancel()
+
+	// We mint some grouped assets to use in the test. These assets are
+	// minted on the default tapd instance that is always created in the
+	// integration test (connected to lnd "Alice").
+	firstBatch := MintAssetsConfirmBatch(
+		t.t, t.lndHarness.Miner.Client, t.tapd,
+		[]*mintrpc.MintAssetRequest{issuableAssets[0]},
+	)[0]
+
+	var (
+		firstBatchGenesis = firstBatch.AssetGenesis
+		aliceTapd         = t.tapd
+		aliceLnd          = t.lndHarness.Alice
+		bobLnd            = t.lndHarness.Bob
+	)
+
+	// We create a second tapd node that will be used to simulate a second
+	// party in the test. This tapd node is connected to lnd "Bob".
+	bobTapd := setupTapdHarness(t.t, t, bobLnd, t.universeServer)
+	defer func() {
+		require.NoError(t.t, bobTapd.stop(!*noDelete))
+	}()
+
+	// And now we prepare the multisig addresses for both levels. On the
+	// BTC level we are going to do a Tapscript based 2-of-2 multisig using
+	// OP_CHECKSIGADD. On the asset level we are going to use a 2-of-2
+	// MuSig2 scheme. The BTC level key is going to be called the "internal
+	// key" and the asset level key is going to be called the "script key".
+	aliceScriptKey, aliceInternalKey := deriveKeys(t.t, aliceTapd)
+	bobScriptKey, bobInternalKey := deriveKeys(t.t, bobTapd)
+
+	// Create the BTC level multisig script, using OP_CHECKSIGADD.
+	btcTapscript, err := txscript.NewScriptBuilder().
+		AddData(schnorr.SerializePubKey(aliceInternalKey.PubKey)).
+		AddOp(txscript.OP_CHECKSIG).
+		AddData(schnorr.SerializePubKey(bobInternalKey.PubKey)).
+		AddOp(txscript.OP_CHECKSIGADD).
+		AddInt64(2).
+		AddOp(txscript.OP_EQUAL).
+		Script()
+	require.NoError(t.t, err)
+	btcTapLeaf := txscript.TapLeaf{
+		LeafVersion: txscript.BaseLeafVersion,
+		Script:      btcTapscript,
+	}
+
+	// The actual internal key of the BTC level Taproot output will be the
+	// provably un-spendable NUMS key.
```

Comment:
are we using NUMS for a security reason, or just to keep things simple? Couldn't the internal key be a alice<>bob combined key?

Reply:
Just for demonstration purposes, to disable the key spend. But yes, since this is a 2-of-2, the internal key could've just been a combined MuSig2 key of the two parties.
But since the demo is for showing how to execute a sibling tap leaf, we constructed it this way.